### PR TITLE
Unify Skin controls

### PR DESCRIPTION
**Pull request by ronso0**
This keeps common skin settings when changing skins.
Continuation of #1642, fixes [Bug 1740513](https://bugs.launchpad.net/mixxx/+bug/1740513)

Are the following COs okay for everyone?
`[Skin],show_4decks`

`[Skin],show_4effectunits`
`[Skin],show_superknobs` (previous [Master],... is used in only one official mapping)

`[Skin],show_spinnies`
`[Skin],show_coverart`
`[Skin],show_big_spinny_coverart`

`[Skin],show_8_hotcues`
`[Skin],show_starrating]`
`[Skin],show_rate_controls`
`[Skin],show_eq_knobs`
`[Skin],show_eq_kill_buttons`
`[Skin],show_xfader`
`[Skin],show_loop_beatjump_controls`

`[Skin],sampler_rows` = number of sampler rows
`[Skin],sampler_row_N` = trigger CO for sampler rows WidgetStack
`[Skin],sampler_row_current/..next/..prev` = COs to cycle throught sampler rows WidgetStack
`[Skin],sampler_row_N_expanded`

### CODE REVIEW

**Comment by daschuer** [comment]
This works for me. Thank you.

**Comment by ronso0** [comment]
Woooo, grep and sed are my friends now.
Please test.

**Comment by Be-ing** [comment]
There is now a merge conflict in res/skins/LateNight/deck_row_5_transportLoopJump.xml.

**Comment by ronso0** [comment]
(rebased)

File: res/skins/Deere (64 Samplers)/skin.xml [hunk 1]
```diff
@@ -8,37 +8,48 @@
     <language>en</language>
     <license>Creative Commons Attribution, Share-Alike 3.0 Unported</license>
     <attributes>
+      <!-- Set skin defaults. -->
       <attribute config_key="[Master],num_decks">4</attribute>
-      <attribute config_key="[Master],num_samplers">64</attribute>
+      <attribute config_key="[Master],num_samplers">16</attribute>
```
Comment (Be-ing) [inline res/skins/Deere (64 Samplers)/skin.xml]: Why did you change this?

File: res/skins/Deere (64 Samplers)/skin.xml [hunk 2]
```diff
@@ -8,38 +8,48 @@
...
-      <attribute persist="true" config_key="[VinylControl],show_vinylcontrol">0</attribute>
-
-      <attribute persist="true" config_key="[Samplers],show_samplers">1</attribute>
-      <attribute persist="true" config_key="[Samplers],show_expanded">0</attribute>
-
-      <attribute persist="true" config_key="[EffectRack1],show">1</attribute>
-      <attribute persist="true" config_key="[Master],show_superknobs">0</attribute>
+      <attribute persist="true" config_key="[Skin],show_eq_knobs">1</attribute>
+      <attribute persist="true" config_key="[Skin],show_eq_kill_buttons">1</attribute>
+      <attribute persist="true" config_key="[Skin],show_xfader">1</attribute>
```
Comment (Be-ing) [inline res/skins/Deere (64 Samplers)/skin.xml]: `[Skin],show_crossfader`? I don't care much either way.

File: res/skins/Deere/skin.xml [hunk 3]
```diff
@@ -8,39 +8,42 @@
     <language>en</language>
     <license>Creative Commons Attribution, Share-Alike 3.0 Unported</license>
     <attributes>
+      <!-- Set skin defaults. -->
       <attribute config_key="[Master],num_decks">4</attribute>
       <attribute config_key="[Master],num_samplers">16</attribute>
       <attribute config_key="[Master],num_preview_decks">1</attribute>
-      <attribute config_key="[Library],show_library">1</attribute>
```
Comment (Be-ing) [inline res/skins/Deere/skin.xml]: Bring this back and use `[Skin],show_library`. I tried starting Mixxx with a new settings directory and Deere did not show the library.

File: res/skins/LateNight/skin.xml [hunk 4]
```diff
@@ -39,26 +39,33 @@
...
+      <!-- Library -->
+      <attribute config_key="[Master],maximize_library">0</attribute>
       <attribute persist="true" config_key="[PreviewDeck],show_previewdeck">0</attribute>
-      <attribute persist="true" config_key="[Master],show_mixer">1</attribute>
-      <attribute persist="true" config_key="[Master],show_eqs">1</attribute>
-      <attribute persist="true" config_key="[Master],show_eq_kill_buttons">1</attribute>
       <attribute persist="true" config_key="[Library],show_coverart">1</attribute>
-      <attribute persist="true" config_key="[Master],show_4decks">0</attribute>
-      <attribute persist="true" config_key="[Master],show_4effectunits">0</attribute>
-      <attribute persist="true" config_key="[Library],show_library">1</attribute>
```
Comment (Be-ing) [inline res/skins/LateNight/skin.xml]: `[Skin],show_library`

File: res/skins/Tango (64 Samplers)/skin.xml [hunk 5]
```diff
@@ -38,61 +38,66 @@
     <attributes>
       <!-- Set skin defaults. -->
       <attribute config_key="[Master],num_decks">4</attribute>
-      <attribute config_key="[Master],num_samplers">64</attribute>
-      <attribute persist="true" config_key="[Master],show_mixer">1</attribute>
-      <attribute persist="true" config_key="[Master],show_eqs">1</attribute>
-      <attribute persist="true" config_key="[Master],show_4decks">0</attribute>
-      <attribute persist="true" config_key="[Master],show_4effectunits">0</attribute>
-      <attribute config_key="[Library],show_library">1</attribute>
```
Comment (Be-ing) [inline res/skins/Tango (64 Samplers)/skin.xml]: `[Skin],show_library`

File: res/skins/Tango/library.xml [hunk 6]
```diff
@@ -102,9 +102,5 @@ Description:
       </Splitter><!-- /Split Tree / Tracks table -->
 
     </Children>
-    <Connection>
-      <ConfigKey persist="true">[Library],show_library</ConfigKey>
```
Comment (Be-ing) [inline res/skins/Tango/library.xml]: `[Skin],show_library`

File: res/skins/Deere (64 Samplers)/skin.xml [hunk 7]
```diff
@@ -8,38 +8,48 @@
     <language>en</language>
     <license>Creative Commons Attribution, Share-Alike 3.0 Unported</license>
     <attributes>
+      <!-- Set skin defaults. -->
       <attribute config_key="[Master],num_decks">4</attribute>
-      <attribute config_key="[Master],num_samplers">64</attribute>
+      <attribute config_key="[Master],num_samplers">16</attribute>
       <attribute config_key="[Master],num_preview_decks">1</attribute>
-      <attribute config_key="[Library],show_library">1</attribute>
```
Comment (Be-ing) [inline res/skins/Deere (64 Samplers)/skin.xml]: `[Skin],show_library`

**Review by Be-ing — CHANGES_REQUESTED** [review]

File: res/skins/Tango (64 Samplers)/skin.xml [hunk 8]
```diff
@@ -38,22 +38,50 @@
     <attributes>
       <!-- Set skin defaults. -->
       <attribute config_key="[Master],num_decks">4</attribute>
-      <attribute config_key="[Master],num_samplers">64</attribute>
-      <attribute persist="true" config_key="[Master],show_mixer">1</attribute>
-      <attribute persist="true" config_key="[Master],show_eqs">1</attribute>
+      <attribute config_key="[Master],num_samplers">16</attribute>
```
Comment (Be-ing) [inline res/skins/Tango (64 Samplers)/skin.xml]: revert back to 64

File: res/skins/Deere (64 Samplers)/skin.xml (hunk 1 above)
Comment (ronso0) [reply res/skins/Deere (64 Samplers)/skin.xml]: copy/paste mistake apparently. thx!

File: res/skins/Deere (64 Samplers)/skin.xml (hunk 2 above)
Comment (ronso0) [reply res/skins/Deere (64 Samplers)/skin.xml]: it was `[Master],show_xfader` in LateNight, that's where I started

File: res/skins/Tango (64 Samplers)/skin.xml (hunk 5 above)
Comment (ronso0) [reply res/skins/Tango (64 Samplers)/skin.xml]: does omitting this CO here affect anything if it isn't used in the here (in this skin)?

File: res/skins/Tango (64 Samplers)/skin.xml (hunk 8 above)
Comment (ronso0) [reply res/skins/Tango (64 Samplers)/skin.xml]: yep. thx!

**Comment by ronso0** [comment]
Neither can I confirm missing waveforms, nor do I see the anylyzing text. Re-analyzing is done around when the skin is fully loaded.

File: res/skins/Tango (64 Samplers)/skin.xml (hunk 5 above)
Comment (Be-ing) [reply res/skins/Tango (64 Samplers)/skin.xml]: I don't think so. But it does need to be added back to Deere.

**Comment by ronso0** [comment]
not finished, yet. I'll also take care of SplitSizes.

**Comment by daschuer** [comment]
Is this really still a 2.2 PR?

**Comment by ronso0** [comment]
> Is this really still a 2.2 PR?

I'd say so.
I think it makes sense to have this in a major 2.2 update where users rather expect/accept to re-configure some skin settings than in a point release.
At which branch would you like to target it?

File: res/skins/Deere (64 Samplers)/skin.xml (hunk 7 above)
Comment (ronso0) [reply res/skins/Deere (64 Samplers)/skin.xml]: not needed in Tango

File: res/skins/Tango (64 Samplers)/skin.xml (hunk 5 above)
Comment (ronso0) [reply res/skins/Tango (64 Samplers)/skin.xml]: added back

File: res/skins/LateNight/skin.xml (hunk 4 above)
Comment (ronso0) [reply res/skins/LateNight/skin.xml]: not needed in LateNight

File: res/skins/Deere/skin.xml (hunk 3 above)
Comment (ronso0) [reply res/skins/Deere/skin.xml]: added back

**Review by Be-ing — CHANGES_REQUESTED** [review]

**Comment by Be-ing** [comment]
8 hotcues setting does not work in LateNight.

I'm ambivalent about whether to merge this to 2.2 or master...

File: res/skins/Deere/skin.xml [hunk 9]
```diff
@@ -8,39 +8,43 @@
...
-      <attribute persist="true" config_key="[Samplers],show_samplers">1</attribute>
-      <attribute persist="true" config_key="[Samplers],show_expanded">0</attribute>
-
-      <attribute persist="true" config_key="[EffectRack1],show">1</attribute>
-      <attribute persist="true" config_key="[Master],show_superknobs">0</attribute>
+      <attribute persist="true" config_key="[Skin],show_eq_knobs">1</attribute>
+      <attribute persist="true" config_key="[Skin],show_eq_kill_buttons">1</attribute>
+      <attribute persist="true" config_key="[Skin],show_xfader">1</attribute>
+      <!-- Effects -->
+      <attribute persist="true" config_key="[EffectRack1],show">0</attribute>
```
[the remark this answers is not in the document]
Comment (ronso0) [reply res/skins/Deere/skin.xml]: fixed.

File: res/skins/Deere/skin.xml [hunk 10]
```diff
@@ -8,39 +8,43 @@
...
       <attribute persist="true" config_key="[Deere],show_track_info">1</attribute>
       <attribute persist="true" config_key="[Deere],show_bpm_info">1</attribute>
-      <attribute persist="true" config_key="[Deere],show_more_hotcues">0</attribute>
-
+      <!-- Mixer -->
       <attribute persist="true" config_key="[Master],show_mixer">1</attribute>
-      <attribute persist="true" config_key="[Deere],show_crossfader">1</attribute>
-      <attribute persist="true" config_key="[Master],show_eqs">1</attribute>
-      <attribute persist="true" config_key="[Master],show_killswitches">0</attribute>
-      <attribute persist="true" config_key="[Master],show_faders">1</attribute>
```
[the remark this answers is not in the document]
Comment (ronso0) [reply res/skins/Deere/skin.xml]: whoopsy.. changed to `[Skin],show_faders`

**Comment by ronso0** [comment]
> 8 hotcues setting does not work in LateNight.

fixed.

**Comment by ronso0** [comment]
> I'm ambivalent about whether to merge this to 2.2 or master...

I'm convinced users will appreciate it when checking out different skins, instead of having to go to each skin's settings menu to restore options. So why wait for 2.3?

**Comment by ronso0** [comment]
(rebased, conflicts fixed)

**Comment by daschuer** [comment]
I have just tested it without any issues, thank you. 
However, I vote for merging this to master. 
We do not know how many custom skins and  mappings are effected out there and a normal beta period would be fair. 
What do others think?

**Comment by Be-ing** [comment]
I'm also wary to rush this for 2.2 considering the potential for regressions. Although we have already identified and fixed some of those, I am concerned there may be more subtle regressions lurking here.

There is a new conflict since #1840 was merged.

**Comment by ronso0** [comment]
> ... and a normal beta period would be fair.

You're right.

**Comment by ronso0** [comment]
Some improvements for Deere:
- use button templates in the skin settings menu
- convert  category labels (MiXER, EFFECTS, ...) to toggles like in LateNight (more compact menu)
- add MIXER toggle to the tool bar
- fixed sampler rows selection

check!
deere_add-mixer-toggle

**Comment by ronso0** [comment]
I'd say we should merge this now before more conflicts pop up.

File: res/skins/Deere/skinsettings_button.xml [hunk 11]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE template>
+<!--
+  Description:
+    A button that has a left-click display control and a right-click control.
```
Comment (Be-ing) [inline res/skins/Deere/skinsettings_button.xml]: This description is incorrect.

File: res/skins/Deere/skinsettings_category_button.xml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE template>
+<!--
+  Description:
+    A button that has a left-click display control and a right-click control.
```
Comment (Be-ing) [inline res/skins/Deere/skinsettings_category_button.xml]: This description is also incorrect.

File: res/skins/Deere/style.qss [hunk 13]
```diff
@@ -113,11 +113,17 @@
 
 #Spacer22 {
   background-color: #222222;
-}
+} /*
```
Comment (Be-ing) [inline res/skins/Deere/style.qss]: delete commented out code

**Review by Be-ing — CHANGES_REQUESTED** [review]

File: res/skins/Deere/style.qss (hunk 13 above)
Comment (ronso0) [reply res/skins/Deere/style.qss]: hmm I'll uncomment it then. It's often useful having such snippets ready for debugging and tests.

File: res/skins/Deere/skinsettings_button.xml (hunk 11 above)
Comment (ronso0) [reply res/skins/Deere/skinsettings_button.xml]: fixed.

**Comment by Be-ing** [comment]
Thanks for your work and sorry to make you rebase so many times for this branch.

**Comment by ronso0** [comment]
never mind, had to iron out some mistakes anyway..

**Comment by nikmartin** [comment]
Just FYI Tango skin loop controls are not visible on master branch, assuming it's this PR.
